stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -14,474 +14,292 @@
+  //  it was causing ambiguities
+  /*template <typename Container,
+            require_t<is_detected<Container, member_size_t>>...,
+            require_t<std::is_arithmetic<value_type_t<Container>>>...,
+            typename... Pargs>
+  size_t count_var_impl(size_t count, const Container& x,
+                        const Pargs&... args) const {
+    return count_var_impl(count, args...);
+    }*/
+
+  template <typename... Pargs>
+  size_t count_var_impl(size_t count, const var& x,
+                        const Pargs&... args) const {
+    return count_var_impl(count + 1, args...);
+  }
+
+  template <typename... Pargs, typename Arith,
+            require_arithmetic_t<scalar_type_t<Arith>>...>
+  size_t count_var_impl(size_t count, Arith& x, const Pargs&... args) const {
+    return count_var_impl(count, args...);
+  }
+
+  size_t count_var_impl(size_t count) const { return count; }
+
+  /**
+   * Count the number of scalars of type T in the input argument list
+   *
+   * @tparam Pargs Types of input arguments
+   * @return Number of scalars of type T in input
+   */
+  template <typename... Pargs>
+  size_t count_var(const Pargs&... args) const {
+    return count_var_impl(0, args...);
+  }
+
+  template <typename... Pargs>
+  void save_varis(vari** dest, const var& x, const Pargs&... args) const {
+    *dest = x.vi_;
+    save_varis(dest + 1, args...);
+  }
+
+  template <typename... Pargs, typename Vec,
+            require_vector_like_vt<is_var, Vec>...>
+  void save_varis(vari** dest, const Vec& x, const Pargs&... args) const {
+    for (size_t i = 0; i < x.size(); ++i) {
+      dest[i] = x[i].vi_;
+    }
+    save_varis(dest + x.size(), args...);
+  }
+
+  template <typename... Pargs, typename Mat,
+            require_t<is_detected<Mat, operator_paren_access_t>>...,
+            require_t<is_var<value_type_t<Mat>>>...>
+  void save_varis(vari** dest, const Mat& x, const Pargs&... args) const {
+    for (size_t i = 0; i < x.size(); ++i) {
+      dest[i] = x(i).vi_;
+    }
+    save_varis(dest + x.size(), args...);
+  }
+
+  template <typename R, require_arithmetic_t<scalar_type_t<R>>...,
+            typename... Pargs>
+  void save_varis(vari** dest, const R& x, const Pargs&... args) const {
+    save_varis(dest, args...);
+  }
+
+  void save_varis(vari**) const {}
+
+  var operator()(const std::vector<M>& vmapped, std::size_t grainsize,
+                 const Args&... args) const {
     const std::size_t num_jobs = vmapped.size();
 
     if (num_jobs == 0)
-      return init;
+      return var(0.0);
+
+    const std::size_t num_sliced_terms = count_var(vmapped);
+    const std::size_t num_shared_terms = count_var(args...);
+
+    vari** varis = ChainableStack::instance_->memalloc_.alloc_array<vari*>(
+        num_sliced_terms + num_shared_terms);
+    double* partials = ChainableStack::instance_->memalloc_.alloc_array<double>(
+        num_sliced_terms + num_shared_terms);
```

Comment:
Should we be using one of the tbb scalable allocators here?

https://software.intel.com/en-us/node/810736

Reply:
Not needed. We can just load the tbbmalloc_proxy library to change all malloc calls to use the scalable allocator.

Right now the real bottleneck is that we do not construct deep copies from pure values, I think. If we create first pure values then we can create deep bar copies from contiguous chunks of value memory which is better for memory locality. At least this is how my intuition goes.

Reply:
> If we create first pure values then we can create deep bar copies from contiguous chunks of value memory which is better for memory locality. At least this is how my intuition goes.

What do you mean pure values here?